db/dump.py

```diff
@@ -112,45 +175,63 @@ def dump_db(location, threads=None, incremental=False, dump_id=None):
-    with open(archive_path, "w") as archive:
-
-        pxz_command = ["pxz", "--compress"]
-        if threads is not None:
-            pxz_command.append("-T %s" % threads)
-        pxz = subprocess.Popen(pxz_command, stdin=subprocess.PIPE, stdout=archive)
-
-        # Creating the archive
-        with tarfile.open(fileobj=pxz.stdin, mode="w|") as tar:
-            # TODO(roman): Get rid of temporary directories and write directly to tar file if that's possible.
-            temp_dir = tempfile.mkdtemp()
-
-            # Adding metadata
-            schema_seq_path = os.path.join(temp_dir, "SCHEMA_SEQUENCE")
-            with open(schema_seq_path, "w") as f:
-                f.write(str(db.SCHEMA_VERSION))
-            tar.add(schema_seq_path,
-                    arcname=os.path.join(archive_name, "SCHEMA_SEQUENCE"))
-            timestamp_path = os.path.join(temp_dir, "TIMESTAMP")
-            with open(timestamp_path, "w") as f:
-                f.write(time_now.isoformat(" "))
-            tar.add(timestamp_path,
-                    arcname=os.path.join(archive_name, "TIMESTAMP"))
-            tar.add(DUMP_LICENSE_FILE_PATH,
-                    arcname=os.path.join(archive_name, "COPYING"))
-
-            archive_tables_dir = os.path.join(temp_dir, "abdump", "abdump")
-            utils.path.create_path(archive_tables_dir)
-            _copy_tables(archive_tables_dir, start_t, end_t)
-            tar.add(archive_tables_dir, arcname=os.path.join(archive_name, "abdump"))
+    _dump_tables(
+        archive_path=archive_path,
+        threads=threads,
+        dataset_dump=False,
+        time_now=time_now,
+        start_t=start_t,
+        end_t=end_t,
+    )
+    return archive_path
 
-            shutil.rmtree(temp_dir)  # Cleanup
 
-        pxz.stdin.close()
+def _copy_table_into_multiple_files(cursor, table_name, query, tar, archive_name):
+    """Copies data from a table into a file within a specified location.
 
-    return archive_path
+    Args:
+        cursor: a psycopg2 cursor
+        table_name: the name of the table to be copied.
+        query: the select query for getting data from the table, with appropriate LIMIT and OFFSET parameters.
+        tar: the TarFile object to which the table dumps must be added
+        archive_name: the name of the archive
+    """
+    location = tempfile.mkdtemp()
+    done = False
+    offset = 0
+    while True:
+        file_name = str(uuid.uuid4())
+        path = os.path.join(location, table_name, file_name[0], file_name[0:2])
+        utils.path.create_path(path)
+        with open(os.path.join(path, file_name), "a") as f:
+            logging.info(" - Copying table {table_name} to {file_name}...".format(table_name=table_name, file_name=file_name))
+            cursor.copy_to(f, query.format(limit=ROWS_PER_FILE, offset=offset))
+            offset += ROWS_PER_FILE
+            if f.tell() == 0:
+                done = True
```

Comment:
I think this would be more explicit if it was called something like `no_more_rows` - I think `f.tell()` will be 0 when the query results in 0 rows because the offset is higher than the number of rows in the table?

Reply:
on it.